### DISCUSSION TITLE
ANN: CoCalc Slides -- create slides in CoCalc for presentations, with Jupyter code and latex math

### DISCUSSION THREAD

**Discussion by williamstein** [sagemathinc/cocalc#6420] [Announcements]
We just launched "CoCalc Slides"!    To try them, click +New --> Slides.

<img width="1609" alt="image" src="https://user-images.githubusercontent.com/1276278/218190855-6dd7a7a8-ab6d-4f70-b8ac-73b641dd418d.png">

You can create any number of slides, and use all the same tools as from the whiteboard: text, notes, freehand drawing, Jupyter code cells, mind maps, icons, and frames.  Of course, text is markdown or rich editable with full latex formula and code block support.

**Comment by mforbes** [comment]
I am just trying slides.  Here is some feedback from someone just using this for the first time.

* Cool!

* When I click  to edit the slide title/subtitle, I am confused as to why there are blank lines before and after the heading?
   ```
   <blank line here>
   #
   <blank line here>
   ```

* When I click to enter a code cell, why are there 6 identical entries 2+3?  Is that somehow related to the size?  Oh! Those cells have different border colors... (can barely see that with my current monitor).  Ah, the 2+3 is example of code.

   I think I get it now: all of the tools open a panel with various different options allowing users to customize their presets.  This might be nice at some point, but as a new user, I expect that if I click the Jupyter Code Cell button, that it will insert a Jupyter Code Cell.  There was quite a bit of stumbling before I realized I had to drag where I wanted the cell to actually get one.  (Similar issue to my experience with pens in the whiteboard - the customization might eventually be useful, but is not what I expected as the first thing to see when trying this out.

* On Chrome at least, some expected keyboard shortcuts do not work.  For example `delete` does not delete selected items, `ctrl-x` or `command-x` (on Mac) do not cut etc.  (I am guessing that this has something to do with the browser cut/paste functionality etc. being active when editing text, but the CoCalc cut/paste menus being active on elements?  It is a little confusing the two different meanings of "Cut".)  Deleting items can be very difficult because the context menu appears to the lower left, and this can easily be off the visible screen with larger shapes etc.  The "Cut" option always works... but there is no delete button.

* How to copy and paste slides?  A common workflow for me is to design a slide, then duplicate it many times and change the content.  (Or duplicate it and change parts of the content so that some parts stay constant as I present.)  I thought maybe this could be done from the overview page or right-clicking on the contents, but I cannot see a way to do it.

* My main concern right now is what I [mentioned in the other discussion](https://github.com/sagemathinc/cocalc/discussions/6427#discussioncomment-5185416): I want to ensure I can work/present offline if needed and I am concerned that if I do too much in the silde right now, I will lock myself in before my flight tomorrow where I want to work on the slides:-)

Let me know if you want me to open any issues.

**Reply by williamstein** [reply]
> [...] Ah, the 2+3 is example of code.

It's exactly the same as with colors, notes, etc.,  It's just a bunch of presets.   I think this model works well for some things, e.g., colors, but much less well for other things, e.g., code cells and also icons, where it is very confusing.   For those, it would be better to just have 2 or 3 presets (at most), and an easy way to add new ones yourself.

>  For example delete does not delete selected items

It actually normally does.   However, there are two selection modes, one green, where your input goes to the editor, and the other blue, where delete, cut, paste work as normal.    This is identical to the whiteboard.  If you have something selected in green edit mode, hit escape to switch to the blue "object select mode".    This makes a lot of sense when editing code and text, but seems weird for a sketched curve.

> but there is no delete button.

I could add that, but it is on the context menu the appears right below the object you're editing.   

> Deleting items can be very difficult because the context menu appears to the lower left, and this can easily be off the visible screen with larger shapes etc. 

I could make it dynamically move so it isn't off the screen.  This can be confusing and less predictable (it's what miro does), but at least you don't have to scroll around to see the menu.  

> How to copy and paste slides? 

You can currently only select everything in a slide, then hit command+c (or ctrl+v), then paste it.  To select everything, draw a selection rectangle around everything with the mouse.    There's no selection implemented in the pages view like one has in powerpoint/google slides, where the entire pages view has a full selection model.  

> I want to ensure I can work/present offline

The canonical solution to that for cocalc is to install cocalc-docker.  It takes a lot of disk space for the full version.  There is also a lite version in case disk space is a concern, but I update that much less frequently.

> Let me know if you want me to open any issues.

Please open issues for any and all significant problems to you that you don't already find in our issue tracker!

**Reply by mforbes** [reply]
> > For example delete does not delete selected items
> 
> It actually normally does. However, there are two selection modes, one green, where your input goes to the editor, and the other blue, where delete, cut, paste work as normal. This is identical to the whiteboard. If you have something selected in green edit mode, hit escape to switch to the blue "object select mode". This makes a lot of sense when editing code and text, but seems weird for a sketched curve.

Aha: so it does.  I kind of noticed this, but was never conscious of it until you just described "green" vs "blue" edit modes here.  Definitely should be described [in the docs](https://doc.cocalc.com/slides.html#cut-or-copy-and-paste) and in a demo/tutorial of how to use slides etc. once things stabilize.  (I did not notice this in your talk, but I skimmed it, so it might have been there.)

> > but there is no delete button.
>
> I could add that, but it is on the context menu the appears right below the object you're editing.

I think I feel the same way... and cut works, so no need to clutter the interface.

> > Deleting items can be very difficult because the context menu appears to the lower left, and this can easily be off the visible screen with larger shapes etc.
>
> I could make it dynamically move so it isn't off the screen. This can be confusing and less predictable (it's what miro does), but at least you don't have to scroll around to see the menu.

I think it is probably worth it for new users... maybe something that could be turned off?

> > How to copy and paste slides?

Added issue: https://github.com/sagemathinc/cocalc/issues/6513

> > I want to ensure I can work/present offline
>
> ...[use] cocalc-docker.

Now that I have a new computer with sufficient disk space, I will try this again... not a possibility before when I was living with a spare 8GB :-/

**Reply by williamstein** [reply]
> Now that I have a new computer with sufficient disk space, I will try this again... 

What computer do you have?

cocalc-docker takes maybe 25GB.  It's not tiny.

It's also possible to run cocalc directly (no vm at all) on linux and mac, but not documented or well supported, but only due to lack of time to do so.  That takes far less disk space, and I've been using it a lot lately (on both macos and linux lately), so it tends to work OK.  There is one big issue though with running Jupyter kernels under cocalc-docker on MacOS directly, which I need to fix.   If I had more resources (i.e., money) having a native cocalc desktop app would be a high priority.

**Reply by mforbes** [reply]
@williamstein MacBook Pro M1 Max

Some more comments after having given a presentation:

* Public sharing (at least link-sharing) does not seem to work well: [UI Colloquium](https://cocalc.com/share/public_paths/3a513f5e3ef8572e49552bce82a974db632126ab).  In particular, output from code cells (plot pngs) are unavailable.  I even tried making the entire folder public.
* I ran into issues using Full Screen on my Wacom One tablet as a secondary monitor.  Specifically, when drawing, the sketches would appear shifted by a large amount from the location of my stylus.  Once I finished my stroke, it would snap back into place correctly, but this made it very hard to write.  Happened with both Chrome and Firefox.  Mirroring my main display worked fine, but I wanted to use my display for notes, a timer etc. but this also does not work because Full Screen mode blackens other displays.  (Would be nice to build these into a presenter display.)

  Here is a demonstration of this issue captured in the default Wacom One resolution (in case the resolution plays a role in the offset_.

  ![CoCalcDrawingIssue](https://user-images.githubusercontent.com/1028433/224459798-1154dcb4-932c-4d3a-9ced-4fdbae5ddc8a.gif)

  I found a workaround using the [Windowed](https://chrome.google.com/webstore/detail/windowed-floating-youtube/gibipneadnbflmkebnmcbgjdkngkbklb) Chrome extension which captures the Full Screen request and allows you to display "Full Screen" in a separate window that can be placed on the tablet.  This worked with two caveats:
  * I could not figure out how to exit full screen mode, so I would have to close the "windowed" window and reopen.
  * The tool menu vanished and I could not figure out how to get it back.  Keyboard shortcuts worked so I could activate pens, select, etc. but could not change colour etc.
  * Drawing is still very sensitive and the output shapes are not very smooth.

  Here is what entering `windowed` mode looks like.  Note how rough the last "circle" is.  This is what I mean by the outputs not being very smooth:
 
  ![CoCalcWindowed](https://user-images.githubusercontent.com/1028433/224459951-78bbd138-f23d-47ca-93f1-37263a5a06f3.gif)
    
  For these reasons, I could not effectively draw on the slides, and I used a neighbouring desktop with Notability for sketches, demonstrating the useful snap-to-circle feature, as well as a smooth circleish object without snap.

  ![CoCalcNotability](https://user-images.githubusercontent.com/1028433/224460098-e2560bf9-8b82-4dab-9038-e5683613fd1b.gif)

* Selecting items while editing the presentation could be painful as often items on top would extend further than they appeared.  Sending objects to the background helps, but is slow -- are there shortcuts for doing this?
* Duplicating slides is still very hard: I tried your suggestion to select everything, then copy/paste in "blue edit" mode.  This works, but when pasting, the objects do not go into the same spot as they were.  To get things to align, I then need to try to move everything many times.  In Keynote, for example, when you copy an object from one slide to another, it initially pastes in the same location.  Subsequent pastes will be displaced (specified by the mouse position) by a constant offset, allowing a grid of objects to be made.  The Keynote behaviour is quite nice in this regard and perhaps could be emulated here?
* Editing long text cells (esp. equations) or long code cells causes them to flow off the slide, which can make editing difficult -- especially if the font size is large.  I would really have preferred to be able to edit these things in a Notebook or a text file.  Looking forward to that conversion.
* Is there a way to Run All cells?
* Dragging and dropping images works reasonably well: is there any way for pasting from a screen capture to work?  (Currently, I have to take the screen cap, save it to file, then drag that file... then delete that file locally.)
* Being able to paste images from the internet would be helpful too: I had to add these by typing `<img src="..." width="100%"/>` each time which was a bit of a pain.
* I do miss being able to copy and past PDF chunks without rasterizing...
* Have not tried embedding movies yet, but will want to in the future.
* I really wanted a few features regarding navigation:
  * The ability in RISE presentations of "sub-slides" which go down instead of across.  These can contain extra material that is easily skipped if needed.
  * A way to skip or hide slides during a presentation.
  * A large-ish whiteboard to the right of each slide for sketching during a presentation... I included blank slides for sketching, but could imaging these might not be big enough.  Maybe the whiteboard could be a sub-slide?  Need to fix the sketching issues first though.
* Is there a way of embedding the Terminal in a slide?  My next talk is about testing in python, and I will need to do some terminal interactions.  (Ideally, I would like to have the commands stored in the slides or a notebook, then have them sent to the terminal to execute.  Have not found a good way to do this yet: can't use `!command` because sometimes I need variables passed from one cell to another, or need to change into a directory.  Similar issues with `%%bash`.)

That is all I can think of right now.  In summary, it worked reasonably well (except that I had not duplicated it on my Docker image... and then almost could not connect to the WiFi before my talk!) but took me longer than I would like to create the slide show with most of the extra time wasted trying to select, edit, and position objects.  (Another reason I would first like to edit everything in something like a notebook, then do pure visual formatting later.)

Thanks!

**Comment by williamstein** [comment]
> Is there a way of embedding the Terminal in a slide? 

There isn't a way to do this yet.  However, I've been very seriously considering making it possible to just embed all document types in whiteboards/slides.  It would be extremely powerful, providing a sort of shared desktop environment that is persistent, etc.  And, with realtime collaboration, thing work fine if you have a document both in the whiteboard/slides *and* open in a tab, unlike with a traditional operating system.

Also, many many thanks for your detailed comments, which I'll get to in due order.

**Comment by mforbes** [comment]
I came back to this (am giving another [Colloquium today at noon on Zoom](https://s3.wp.wsu.edu/uploads/sites/1777/2023/03/Forbes-033023.pdf)) and ran into some more difficulties.  In particular, I found it very difficult to move some objects as shown below.  Even when selecting just a single object, moving it would move the text underneath too.

![CoCalcMovingBug](https://user-images.githubusercontent.com/1028433/228884619-932e1ff8-86cc-42a6-9aee-afee9894049b.gif)

After this, I noticed that objects on many slides had all moved down a similar amount, and had to correct them all.  Perhaps somehow the selection went through multiple slides and objects?  No time right now to try to reproduce as I am trying to clean everything up for the final presentation:-/
